File: backend/internal/budgets/repository.go
package budgets

import (
	"context"
	"database/sql"
	"fmt"
	"log"
)

type budgetRepository struct {
	db *sql.DB
}

func NewBudgetrepository(db *sql.DB) budgetRepository {
	return budgetRepository{db: db}
}

func (b budgetRepository) GetBuget(ctx context.Context) (Budget, error) {
	q := `SELECT * FROM budgets`

	budget := Budget{}
	err := b.db.QueryRowContext(ctx, q).Scan(&budget.ID, &budget.Amount, &budget.Currency, &budget.CreatedAt, &budget.UpdatedAt)
	if err != nil {
		if err == sql.ErrNoRows {
			return Budget{}, fmt.Errorf("%w", err)
		}
		return Budget{}, fmt.Errorf("%w", err)
	}
	log.Printf("get budget query failed: %v", err)
	return budget, nil
}

func (b *budgetRepository) CreateBudget(ctx context.Context, budget *Budget) (Budget, error) {
	q := `INSERT INTO budgets (id,amount, currency)
	VALUES (
		gen_random_uuid(),
		$1,
		$2
	)
	RETURNING id, amount, currency, created_at, updated_at
	`
	err := b.db.QueryRowContext(ctx, q, budget.Amount, budget.Currency).Scan(&budget.ID, &budget.Amount, &budget.Currency, &budget.CreatedAt, &budget.UpdatedAt)
	if err != nil {
		return Budget{}, fmt.Errorf("%w", err)
	}
	log.Printf("create budget query failed: %v", err)
	return *budget, nil
}

func (b *budgetRepository) UpdateBudget(ctx context.Context, budget Budget) (Budget, error) {
	q := `UPDATE budgets
	SET amount = $1, currency = $2
	WHERE id = $3
	RETURNING id, amount, currency, created_at, updated_at
	`
	err := b.db.QueryRowContext(ctx, q, budget.Amount, budget.Currency, budget.ID).Scan(&budget.ID, &budget.Amount, &budget.Currency, &budget.CreatedAt, &budget.UpdatedAt)
	if err != nil {
		if err == sql.ErrNoRows {
			return Budget{}, fmt.Errorf("%w", err)
		}
		return Budget{}, fmt.Errorf("%w", err)
	}

	log.Printf("update budget query failed: %v", err)
	return budget, nil
}

func (b *budgetRepository) BudgetExist(ctx context.Context) (bool, error) {
	var exist bool
	q := `SELECT EXISTS(SELECT 1 FROM BUDGETS)`
	err := b.db.QueryRowContext(ctx, q).Scan(&exist)
	if err != nil {
		return exist, err
	}
	log.Printf("budget exist query failed: %v", err)
	return exist, nil
}

File: mobile/features/budgets/repository.ts
import type * as SQLite from 'expo-sqlite';

import type { Budget, BudgetPayload } from '@/features/budgets/types';

type BudgetRow = {
  id: string;
  amount: number;
  currency: string;
  period_type: Budget['periodType'];
  period_days: number | null;
  created_at: string;
  updated_at: string;
};

function rowToBudget(row: BudgetRow): Budget {
  return {
    id: row.id,
    amount: row.amount,
    currency: row.currency,
    periodType: row.period_type,
    periodDays: row.period_days,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

function createBudgetId(): string {
  return `${Date.now()}-${Math.random().toString(36).slice(2, 11)}`;
}

export async function hasCompletedOnboarding(
  db: SQLite.SQLiteDatabase,
): Promise<boolean> {
  const row = await db.getFirstAsync<{ onboarding_completed_at: string | null }>(
    'SELECT onboarding_completed_at FROM app_meta WHERE id = 1',
  );
  return row?.onboarding_completed_at != null;
}

export async function createBudget(
  db: SQLite.SQLiteDatabase,
  payload: BudgetPayload,
): Promise<Budget> {
  const now = new Date().toISOString();
  const id = createBudgetId();
  const periodDays =
    payload.periodType === 'custom' ? (payload.periodDays ?? null) : null;

  await db.runAsync(
    `INSERT INTO budgets (id, amount, currency, period_type, period_days, created_at, updated_at)
     VALUES (?, ?, ?, ?, ?, ?, ?)`,
    id,
    payload.amount,
    payload.currency,
    payload.periodType,
    periodDays,
    now,
    now,
  );

  const row = await db.getFirstAsync<BudgetRow>(
    'SELECT * FROM budgets WHERE id = ?',
    id,
  );

  if (!row) {
    throw new Error('Failed to create budget');
  }

  return rowToBudget(row);
}

export async function hasSavedBudget(
  db: SQLite.SQLiteDatabase,
): Promise<boolean> {
  const row = await db.getFirstAsync<{ count: number }>(
    'SELECT COUNT(*) AS count FROM budgets',
  );
  return (row?.count ?? 0) > 0;
}

export async function markOnboardingComplete(
  db: SQLite.SQLiteDatabase,
): Promise<void> {
  const now = new Date().toISOString();
  await db.runAsync(
    'UPDATE app_meta SET onboarding_completed_at = ? WHERE id = 1',
    now,
  );
}

export async function completeOnboarding(
  db: SQLite.SQLiteDatabase,
  payload: BudgetPayload,
): Promise<Budget> {
  let budget: Budget | null = null;

  await db.withTransactionAsync(async () => {
    budget = await createBudget(db, payload);
    await markOnboardingComplete(db);
  });

  if (!budget) {
    throw new Error('Failed to complete onboarding');
  }

  return budget;
}
